src/Components/Web.JS/src/NavigationLock.ts

```diff
@@ -0,0 +1,37 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+let numLocksWithPromptEnabled = 0;
+
+export const NavigationLock = {
+  enableNavigationPrompt,
+  disableNavigationPrompt,
+};
+
+function onBeforeUnload(event: BeforeUnloadEvent) {
+  event.preventDefault();
+  // Modern browsers display a confirmation prompt when returnValue is some value other than
+  // null or undefined.
+  // See: https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#compatibility_notes
+  event.returnValue = true;
+}
+
+function enableNavigationPrompt() {
+  if (numLocksWithPromptEnabled === 0) {
+    window.addEventListener('beforeunload', onBeforeUnload);
+  }
+
+  numLocksWithPromptEnabled++;
```

Comment:
Not essential, but instead of reference counting, it might be simpler or more robust to have each NavigationLock instance make up a GUID for itself (or some other ID) and send that to the JS side which can track all of the ones with an active lock in a `Map`. Then we don't have to be concerned about double-registration or double-unregistration.